### ISSUE TITLE
error inserting date-time into Microsoft SQL Server

### ISSUE THREAD

**Issue by az0** [sqlobject/sqlobject#114]
**tldr: Microsoft SQL Server does not like more than three decimal places**

After installing my working program on a new machine and updating SQLObject and adodbapi, I can no longer insert date-time values.  

The working system is Windows 7, Python 2.7.9 32-bit, sqlobject 1.7.0b1dev-r4758, adodbapi 2.6.0.7

The problem system is Windows Server 2012,  Python 2.7.10 32-bit, SQLObject 2.1.2, adodbapi 2.6.0.7.

Downgrading from SQLObject 2.1.2 to SQLObject 1.7.5 helps

This is the traceback
````
File "C:\Python27\lib\site-packages\sqlobject\main.py", line 1231, in __init__
    self._create(id, **kw)
  File "C:\Python27\lib\site-packages\sqlobject\main.py", line 1279, in _create
    self._SO_finishCreate(id)
  File "C:\Python27\lib\site-packages\sqlobject\main.py", line 1304, in _SO_finishCreate
    id, names, values)
  File "C:\Python27\lib\site-packages\sqlobject\dbconnection.py", line 468, in queryInsertID
    return self._runWithConnection(self._queryInsertID, soInstance, id, names, values)
  File "C:\Python27\lib\site-packages\sqlobject\dbconnection.py", line 327, in _runWithConnection
    val = meth(conn, *args)
  File "C:\Python27\lib\site-packages\sqlobject\mssql\mssqlconnection.py", line 158, in _queryInsertID
    c.execute(q)
  File "C:\Python27\lib\site-packages\adodbapi\adodbapi.py", line 871, in execute
    self._execute_command()
  File "C:\Python27\lib\site-packages\adodbapi\adodbapi.py", line 688, in _execute_command
    self._raiseCursorError(klass, _message)
  File "C:\Python27\lib\site-packages\adodbapi\adodbapi.py", line 561, in _raiseCursorError
    eh(self.connection, self, errorclass, errorvalue)
  File "C:\Python27\lib\site-packages\adodbapi\apibase.py", line 53, in standardErrorHandler
    raise errorclass(errorvalue)
adodbapi.apibase.DatabaseError: (-2147352567, 'Exception occurred.', (0, u'Microsoft OLE DB Provider for SQL Server', u'Conversion failed when converting date a nd/or time from character string.', None, 0, -2147217913), None)
Command:
INSERT INTO timeline_tweet (id, user_id, user_screen_name, created_at, text, source, in_reply_to_status_id, in_reply_to_user_id, in_reply_to_screen_name, timeline_id) VALUES (608339172902977537,8983712, N'compassion', '2015-06-09 18:25:35.000000', N'@JeffBoutin We''re so glad you enjoy this experience! Thanks for spreading the word! -Emily', N'<a href="http://www.hootsuite.com" rel="nofollow">Hootsuite</a>', 608253973452222464, 2204890032, N'JeffBoutin', 1)
````
Both are connected to the same remote Microsoft SQL Server version 11.0 SP1 (2012)

When using Microsoft SQL Server Management Studio Express to run this command, it works
````select {ts '2015-06-09 18:25:35.000'}````

But this command (more similar to what SQLObject makes) fails:
````select {ts '2015-06-09 18:25:35.0000'}````

**Comment by phdru** [comment]
Yes, versions >=2.0 pass microseconds with format '%f'. You can try to change the format to avoid microseconds completely:
    DateTimeCol(datetimeFormat='%Y-%m-%d %H:%M:%S')

I'd like to see CREATE TABLE command for the table — what was the format for date/time? Can it be extended to support 6-digits microseconds? Something like
    ALTER TABLE name MODIFY COLUMN col TIME(6)
in MySQL…

**Comment by az0** [comment]
The CREATE TABLE command exported from Microsoft SQL Server Management Studio Express
````
USE [usa_support]
GO

/****** Object:  Table [dbo].[timeline_tweet]    Script Date: 6/9/2015 1:49:01 PM ******/
SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO

CREATE TABLE [dbo].[timeline_tweet](
	[id] [bigint] NULL,
	[user_id] [bigint] NULL,
	[user_screen_name] [nvarchar](20) NULL,
	[created_at] [datetime] NULL,
	[text] [nvarchar](max) NULL,
	[source] [nvarchar](max) NULL,
	[in_reply_to_status_id] [bigint] NULL,
	[in_reply_to_user_id] [bigint] NULL,
	[in_reply_to_screen_name] [nvarchar](20) NULL,
	[timeline_id] [int] NULL,
UNIQUE NONCLUSTERED 
(
	[id] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, FILLFACTOR = 65) ON [PRIMARY]
) ON [PRIMARY] TEXTIMAGE_ON [PRIMARY]

GO

ALTER TABLE [dbo].[timeline_tweet]  WITH CHECK ADD FOREIGN KEY([timeline_id])
REFERENCES [dbo].[timeline] ([id])
GO


````

Using **datetime2** works
````
CREATE TABLE [dbo].[zzz_ahz_test_datetime2](
	[created_at] [datetime2] NULL
	)

 insert into zzz_ahz_test_datetime2 values ('2015-06-09 18:25:35.000000')
````

**Comment by phdru** [comment]
Aha, so the types for DateTime and Time columns with microseconds on MSSQL must be `datetime2(6)` and `time(6)`.

SQLObject cannot fix existing columns, you have to run necessary `ALTER TABLE` yourself. Please report what commands work — I'll add them to the docs.

I'll also think about changing the code to produce proper column types on table creation.

**Comment by az0** [comment]
Normally datetime2 seems like a good choice, but for my purpose I would like to avoid datetime2 because of the other system that reads this, so I changed the definition like this

````
class TimelineTweet(SQLObject):
    user_id = IntCol()
    user_screen_name = UnicodeCol(length=20)
    created_at = DateTimeCol(datetimeFormat='%Y-%m-%d %H:%M:%S')
    [...snip..]
````

And then I upgraded my old machine from sqlobject 1.7.0b1dev-r4758 to 2.1.2, but it seems to be ignoring the datetimeFormat
````
C:\Python27\Scripts>"z:\etl\twitter\tweets2sql.py" -c "mssql://db/db?sspi=1&debug=t" --user WorldVision
 1/QueryAll:  SELECT name FROM sysobjects WHERE type='U'
 1/QueryR  :  SELECT name FROM sysobjects WHERE type='U'
 1/COMMIT  :  auto
 1/QueryAll:  SELECT name FROM sysobjects WHERE type='U'
 1/QueryR  :  SELECT name FROM sysobjects WHERE type='U'
 1/COMMIT  :  auto
 1/QueryAll:  SELECT name FROM sysobjects WHERE type='U'
 1/QueryR  :  SELECT name FROM sysobjects WHERE type='U'
 1/COMMIT  :  auto
 1/QueryAll:  SELECT name FROM sysobjects WHERE type='U'
 1/QueryR  :  SELECT name FROM sysobjects WHERE type='U'
 1/COMMIT  :  auto
*** Archiving user timelime: WorldVision
 1/QueryOne:  SELECT COUNT(*) FROM timeline WHERE (lower(timeline.screen_name) L
IKE ('worldvision'))
 1/QueryR  :  SELECT COUNT(*) FROM timeline WHERE (lower(timeline.screen_name) L
IKE ('worldvision'))
 1/COMMIT  :  auto
 1/Select  :  SELECT TOP 1 timeline.id, timeline.screen_name, timeline.since_id
FROM timeline WHERE (lower(timeline.screen_name) LIKE ('worldvision'))
 1/QueryR  :  SELECT TOP 1 timeline.id, timeline.screen_name, timeline.since_id
FROM timeline WHERE (lower(timeline.screen_name) LIKE ('worldvision'))
 1/COMMIT  :  auto
TimelineArchiver.query({'count': 200, 'screen_name': 'WorldVision', 'since_id':
607926308895629312L})
 1/QueryOne:  SELECT user_id, user_screen_name, created_at, text, source, in_reply_to_status_id, in_reply_to_user_id, in_reply_to_screen_name, timeline_id FROM timeline_tweet WHERE ((timeline_tweet.id) = (608329957257920512))
 1/QueryR  :  SELECT user_id, user_screen_name, created_at, text, source, in_reply_to_status_id, in_reply_to_user_id, in_reply_to_screen_name, timeline_id FROM timeline_tweet WHERE ((timeline_tweet.id) = (608329957257920512))
 1/COMMIT  :  auto
 1/QueryIns:  INSERT INTO timeline_tweet (id, user_id, user_screen_name, created_at, text, source, in_reply_to_status_id, in_reply_to_user_id, in_reply_to_screen_name, timeline_id) VALUES (608329957257920512, 14086764, N\'WorldVision\', \'2015-06-09 17:48:58.000000\', N\'RT @RichStearns: WV staff in SSudan showed me how to measure the upper arm of a starving child. I\'\'ve never seen hunger like this http://t.c\u2026\', N\'<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>\', NULL, NULL, NULL, 26)
 1/COMMIT  :  auto
Traceback (most recent call last):
  File "z:\etl\twitter\tweets2sql.py", line 376, in <module>
    main()
  File "z:\etl\twitter\tweets2sql.py", line 373, in main
    archive_loop(ta)
  File "z:\etl\twitter\tweets2sql.py", line 258, in archive_loop
    archiver.query()
  File "z:\etl\twitter\tweets2sql.py", line 230, in query
    TimelineTweet(**kwargs)
  File "C:\Python27\lib\site-packages\sqlobject\main.py", line 1231, in __init__

    self._create(id, **kw)
  File "C:\Python27\lib\site-packages\sqlobject\main.py", line 1279, in _create
    self._SO_finishCreate(id)
  File "C:\Python27\lib\site-packages\sqlobject\main.py", line 1304, in _SO_finishCreate
    id, names, values)
  File "C:\Python27\lib\site-packages\sqlobject\dbconnection.py", line 468, in queryInsertID
    return self._runWithConnection(self._queryInsertID, soInstance, id, names, values)
  File "C:\Python27\lib\site-packages\sqlobject\dbconnection.py", line 327, in _runWithConnection
    val = meth(conn, *args)
  File "C:\Python27\lib\site-packages\sqlobject\mssql\mssqlconnection.py", line 158, in _queryInsertID
    c.execute(q)
  File "build\bdist.win32\egg\adodbapi\adodbapi.py", line 871, in execute
  File "build\bdist.win32\egg\adodbapi\adodbapi.py", line 688, in _execute_command
  File "build\bdist.win32\egg\adodbapi\adodbapi.py", line 561, in _raiseCursorError
  File "build\bdist.win32\egg\adodbapi\apibase.py", line 53, in standardErrorHandler
adodbapi.apibase.DatabaseError
````

**Comment by phdru** [comment]
Yes, sorry. There is also standard python's datetime converter that doesn't use that format. To change the converter use the following code:

```python
from datetime import datetime
from sqlobject.converters import registerConverter

def DateTimeConverter(value, db):
    return "'%04d-%02d-%02d %02d:%02d:%02d'" % (
        value.year, value.month, value.day,
        value.hour, value.minute, value.second)

registerConverter(datetime, DateTimeConverter)
```

**Comment by phdru** [comment]
datetim2(6): commit b3381c4.

Now I'm thinking about a global switch to turn microseconds off. Something like `sqlobject.use_microseconds(True/False)`.

**Comment by phdru** [comment]
At ed8a18d I add a function col.use_microseconds(True/False). Default is to use microseconds (True). Will be in the next release 2.2.0b1.

**Comment by az0** [comment]
Thanks.  :)

**Comment by phdru** [comment]
I've just released 2.2.0b1. Please test and report if it works and suites your needs.
